Rework HTTP API

jerben opened the pull request.

aronerben commented on sihl/core/ctx.ml:
```diff
@@ -12,3 +12,15 @@ let find key ctx = Hmap.find key ctx.map
+  | None ->
+    counter := !counter + 1;
+    (* Increase counter and calculate md5 hash for consistent id length *)
+    let id = Digest.string (string_of_int !counter) in
```
Consider using something faster

jerben replied:
Fixed.

aronerben replied:
What was changed? I don't see it...

jerben replied:
Fixed pushed :)

aronerben commented on sihl/core/ctx.ml:
```diff
@@ -12,3 +12,15 @@ let find key ctx = Hmap.find key ctx.map
+let sexp_of_t t = Sexplib.Std.sexp_of_string t.id
+let counter = ref 0
+
+let create ?id () =
```
Why can `id` be set here?

jerben replied:
Because the HTTP layer have to be able to set it

aronerben commented on sihl/http/request.ml:
```diff
@@ -0,0 +1,374 @@
+  Opium_kernel.Hmap.Key.create ("id", Sexplib.Std.sexp_of_string)
+;;
+
+let make
```
Allow this to take `~meth` so any type of request can be given.

jerben replied:
I removed `make` since it is tricky to allow polymorphic variants without enumerating all.

jerben replied:
I now still added `meth`.

aronerben commented on sihl/http/request.ml:
```diff
@@ -0,0 +1,374 @@
+  | _ -> None
+;;
+
+let param req key =
```
Check if we need these 5 params

jerben replied:
Fixed.

jerben replied:
I leave it for now, added proper documentation.

aronerben commented on sihl/http/service.ml:
```diff
@@ -0,0 +1,101 @@
+let prefix_route prefix (meth, path, handler) =
+  let path =
+    String.split_on_char '/' path
+    |> List.filter (fun e -> e != "")
```
Only filter left empty string, not right

jerben replied:
Fixed.

aronerben reviewed (commented):
Please add tests for everything

aronerben commented on test/service-tests/cases/session.ml:
```diff
@@ -1,39 +1,36 @@
+    let handler req =
+      let session = Sihl.Middleware.Session.find req in
+      let* () = SessionService.set ctx session ~key:"foo" ~value:"bar" in
+      Lwt.return @@ Sihl.Http.Response.of_plain_text ""
```
Make this return `create ()`

jerben replied:
Fixed.

aronerben reviewed (changes requested):
Get to work, boy.

jerben commented on sihl/middleware/middleware_csrf.ml:
```diff
@@ -1,11 +1,29 @@
+  | _ -> None
+;;
+
+let set session req =
```
Fixed.

jerben commented on sihl/session/dune:
```diff
@@ -6,4 +6,4 @@
  (preprocess
   (pps ppx_fields_conv ppx_deriving_yojson ppx_deriving.eq ppx_deriving.show
-    ppx_deriving.make)))
+    ppx_deriving.make ppx_sexp_conv)))
```
Fixed.

jerben commented on test/service-tests/cases/csrf.ml:
```diff
@@ -1,67 +1,68 @@
+      let token_value = Sihl.Middleware.Csrf.find req in
       Alcotest.(check bool "Has CSRF token" true (not @@ String.equal "" token_value));
-      Lwt.return @@ Sihl.Web.Res.html
+      Lwt.return @@ Sihl.Http.Response.of_plain_text ""
```
Fixed.

jerben commented on test/service-tests/cases/csrf.ml:
```diff
@@ -1,67 +1,68 @@
+    let ctx = Sihl.Http.Request.to_ctx req in
     let* () = Sihl.Repository.Service.clean_all ctx in
-    let handler _ = Lwt.return @@ Sihl.Web.Res.html in
+    let handler _ = Lwt.return @@ Sihl.Http.Response.of_plain_text "" in
```
Fixed.

jerben commented on test/service-tests/cases/csrf.ml:
```diff
@@ -1,67 +1,68 @@
-      Lwt.return Sihl.Web.Res.html
+    let handler tkn req =
+      tkn := Sihl.Middleware.Csrf.find req;
+      Lwt.return @@ Sihl.Http.Response.of_plain_text ""
```
Fixed.

jerben commented on test/service-tests/cases/csrf.ml:
```diff
@@ -96,16 +96,15 @@ struct
+      let token_value = Sihl.Middleware.Csrf.find req in
       Alcotest.(check bool "Has CSRF token" true (not @@ String.equal "" token_value));
-      Lwt.return @@ Sihl.Web.Res.html
+      Lwt.return @@ Sihl.Http.Response.of_plain_text ""
```
Fixed.

jerben commented on test/service-tests/cases/csrf.ml:
```diff
@@ -1,67 +1,68 @@
+    let status = Sihl.Http.Response.status response in
     Alcotest.(check int "Has status 200" 200 status);
     (* Do GET with session to get token *)
     let get_req =
```
Fixed.

jerben commented on test/service-tests/cases/csrf.ml:
```diff
@@ -96,16 +96,15 @@ struct
         ~body:(Cohttp_lwt.Body.of_string "")
```
Fixed.

jerben commented on test/service-tests/cases/csrf.ml:
```diff
@@ -130,14 +129,14 @@ struct
         ~body:(Cohttp_lwt.Body.of_string "csrf=invalid_token")
```
Fixed.

jerben commented on test/service-tests/cases/csrf.ml:
```diff
@@ -114,12 +113,12 @@ struct
+    let ctx = Sihl.Http.Request.to_ctx post_req in
     let* () = Sihl.Repository.Service.clean_all ctx in
-    let handler _ = Lwt.return @@ Sihl.Web.Res.html in
+    let handler _ = Lwt.return @@ Sihl.Http.Response.of_plain_text "" in
```
Fixed.

jerben commented on test/service-tests/cases/csrf.ml:
```diff
@@ -130,14 +129,14 @@ struct
+    let ctx = Sihl.Http.Request.to_ctx post_req in
     let* () = Sihl.Repository.Service.clean_all ctx in
-    let handler _ = Lwt.return @@ Sihl.Web.Res.html in
+    let handler _ = Lwt.return @@ Sihl.Http.Response.of_plain_text "" in
```
Fixed.

jerben commented on test/service-tests/cases/csrf.ml:
```diff
@@ -150,42 +149,41 @@ struct
+    let ctx = Sihl.Http.Request.to_ctx post_req in
     let* () = Sihl.Repository.Service.clean_all ctx in
-    let handler _ = Lwt.return @@ Sihl.Web.Res.html in
+    let handler _ = Lwt.return @@ Sihl.Http.Response.of_plain_text "" in
```
Fixed.

jerben commented on test/service-tests/cases/csrf.ml:
```diff
@@ -150,42 +149,41 @@ struct
-      Lwt.return Sihl.Web.Res.html
+    let handler req =
+      token_ref := Sihl.Middleware.Csrf.find req;
+      Lwt.return @@ Sihl.Http.Response.of_plain_text ""
```
Fixed.

jerben commented on test/service-tests/cases/csrf.ml:
```diff
@@ -150,42 +149,41 @@ struct
     in
-    let ctx = Sihl.Core.Ctx.empty |> Sihl.Web.Req.add_to_ctx post_req in
-    let handler _ = Lwt.return @@ Sihl.Web.Res.html in
+    let handler _ = Lwt.return @@ Sihl.Http.Response.of_plain_text "" in
```
Fixed.

jerben commented on test/service-tests/cases/csrf.ml:
```diff
@@ -198,34 +196,36 @@ struct
+    let ctx = Sihl.Http.Request.to_ctx post_req in
     let* () = Sihl.Repository.Service.clean_all ctx in
-    let handler _ = Lwt.return @@ Sihl.Web.Res.html in
+    let handler _ = Lwt.return @@ Sihl.Http.Response.of_plain_text "" in
```
Fixed.

jerben commented on test/service-tests/cases/csrf.ml:
```diff
@@ -198,34 +196,36 @@ struct
+    let handler req =
+      session_req := Some req;
+      token_ref := Sihl.Middleware.Csrf.find req;
+      Lwt.return @@ Sihl.Http.Response.of_plain_text ""
```
Fixed.

jerben commented on test/service-tests/cases/csrf.ml:
```diff
@@ -150,42 +149,41 @@ struct
         ~body:(Cohttp_lwt.Body.of_string "csrf=aGVsbG8=")
```
Fixed.

jerben commented on test/service-tests/cases/csrf.ml:
```diff
@@ -150,42 +149,41 @@ struct
-    let* _ = wrapped_handler ctx in
+    let* _ = wrapped_handler req in
     (* New request with new session but using old token *)
     let post_req =
```
Fixed.

jerben commented on test/service-tests/cases/csrf.ml:
```diff
@@ -237,14 +237,13 @@ struct
            ~meth:`POST
```
Fixed.

jerben commented:
I pushed my fixes, tests are still missing.

aronerben commented:
@jerben Reviewed

tmattio commented:
Hey @jerben!

Just saw this PR, I'm glad you're liking the new Opium's API.

Were there any API changes that you made to the original API? I've been using it for a month now and I'm pretty happy with it, but I'll happily take your suggestions into account 🙂

jerben commented:
Hey @tmattio

I added some helpers regarding params 

```ocaml
val params3 : t -> string -> string -> string -> (string * string * string) option
```

but I am not sure whether we keep it like that. We plan to add a few helpers on top of `request.mli` using https://github.com/oxidizing/conformist to be able to declaratively define requests by defining schemas for params, query strings and urlencoded body. This could also be used to generate HTTP documentation.

In the next couple of days we are updating the Sihl version in a project to use the new HTTP API, I will let you know how it goes. It is definitely an improvement from what we had before in terms of simplicity and it covers all our needs. :smile: 

Have you thought about web sockets? We might have to touch a few things once we implement that.

I went ahead and implemented one of your proposed router APIs as well. Once httpaf based Opium stabilizes, the plan is to get rid of the compatibility layer.

tmattio commented:
Thanks for the details @jerben 😄 

> added some helpers regarding params

I've hesitated to have `param*` in `Request` or keep them in `Router` and decided to keep them as is, but I'm happy to reconsider if you think that's better to have every utility in `Request` / `Response`!

> I went ahead and implemented one of your proposed router APIs as well

Which one was that? We are currently discussing the new router API, so I'm curious which one you ended up choosing!

> Have you thought about web sockets

On our roadmap, but I didn't work on it yet

> We plan to add a few helpers on top of request.mli using https://github.com/oxidizing/conformist to be able to declaratively define requests by defining schemas for params

That seems really promising! Any thoughts on integrating it with the router directly? You might not want to couple the handler with the route, but in practice, I think there's almost always a 1-to-1 relationship.

Also, do you think that's something you'd like to see in Opium directly or is it too high level for Opium to be concerned about it?

> In the next couple of days we are updating the Sihl version in a project to use the new HTTP API, I will let you know how it goes

That's great, thanks a lot!

jerben commented:
Hey @tmattio 

So I updated a project with about 100 routes and handlers with the new HTTP API, these are my thoughts.

> I've hesitated to have `param*` in `Request` or keep them in `Router` and decided to keep them as is, but I'm happy to reconsider if you think that's better to have every utility in `Request` / `Response`!

I don't think those helpers as it can just be done with an easy pattern match. 

> Which one was that? We are currently discussing the new router API, so I'm curious which one you ended up choosing!

```ocaml
let all =
  [ router ~scope:"" ~middlewares:Middleware.site_middlewares public
  ; router
      ~scope:"/applicants"
      ~middlewares:Middleware.logged_in_site_middlewares
      applicant
  ; router
      ~scope:"/applications"
      ~middlewares:Middleware.applications_site_middlewares
      application
  ; router ~scope:"/admin" ~middlewares:Middleware.admin_middlewares admin
  ; router ~scope:"/api" ~middlewares:[] api
  ]
;;
```
I took this one because I just needed something that worked, but in the long term we probably want something that composes better, so a router that takes routers. Which one is is your current favorite?

> That seems really promising! Any thoughts on integrating it with the router directly? You might not want to couple the handler with the route, but in practice, I think there's almost always a 1-to-1 relationship.

I agree, it should go into the routes/router. I haven't found an API yet that is composable and type safe enough.

> Also, do you think that's something you'd like to see in Opium directly or is it too high level for Opium to be concerned about it?

Yes! Once I played with it enough I would love to open a PR to discuss. Fastify is a similar library and also has built-in support for JSON schema: https://github.com/fastify/fastify/blob/master/docs/Routes.md#shorthand-declaration.

> > In the next couple of days we are updating the Sihl version in a project to use the new HTTP API, I will let you know how it goes

I haven't found anything that's missing, the API seems complete and is quite pleasant to work with!
